src/kernel/spinlock.c

```diff
@@ -0,0 +1,36 @@
+#include "hartlock.h"
+#include "panic.h"
+#include "spinlock.h"
+
+/* Returns 1 if thread is holding the specified spinlock */
+u32 holding(struct spinlock *sp) {
+	struct hart_locals *hart = get_hart_locals();
+	return ((sp->state == 1) && (sp->hart == hart));
+}
+
+void initlock(struct spinlock *sp, const char *name) {
+	sp->name = name;
+	sp->state = 0;
+	sp->hart = 0;
+}
+
+void acquire(struct spinlock *sp) {
+	hart_lock();  // Lock execution to current hart
+	if (holding(sp)) {
+		panic("acquire: {name}", sp->name);
```

Comment:
should probably be `{cstr}`; maybe also a better error message? (`spinlock_lock: tried to acquire {cstr}, but already held it`?)